src/core/externalstorage/qgsexternalstorage.h

```diff
@@ -0,0 +1,190 @@
+/***************************************************************************
+  qgsexternalstorage.h
+  --------------------------------------
+  Date                 : March 2021
+  Copyright            : (C) 2021 by Julien Cabieces
+  Email                : julien dot cabieces at oslandia dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSEXTERNALSTORAGE_H
+#define QGSEXTERNALSTORAGE_H
+
+#include "qgis_core.h"
+#include "qgis_sip.h"
+#include "qgis.h"
+
+#include <QObject>
+#include <QString>
+
+class QgsExternalStorageFetchedContent;
+class QgsExternalStorageStoredContent;
+
+/**
+ * \ingroup core
+ * \brief Abstract interface for external storage - to be implemented by various backends
+ * and registered in QgsExternalStorageRegistry.
+ *
+ * \since QGIS 3.22
+ */
+class CORE_EXPORT QgsExternalStorage
+{
+  public:
+
+    /**
+     * Destructor
+     */
+    virtual ~QgsExternalStorage() = default;
+
+    /**
+     * Unique identifier of the external storage type.
+     */
+    virtual QString type() const = 0;
+
+    /**
+     * Stores file \a filePath to the \a url for this project external storage.
+     * Storing process is run in background.
+     * Returns a QgsExternalStorageStoredContent to follow the status of the stored resource.
+     *
+     * After using this method, user should check if the returned content is not already finished
+     * (storing could be instantaneous, if file has already been stored for instance) and then wait
+     * for QgsExternalStorageStoredContent::stored(), QgsExternalStorageStoredContent::errorOccurred() or
+     * QgsExternalStorageStoredContent::canceled() signals.
+     *
+     * It's possible to give \a authcfg authentication configuration id in case its needed.
+     *
+     * Caller takes ownership of the returned symbol.
+     */
+    virtual QgsExternalStorageStoredContent *store( const QString &filePath, const QString &url, const QString &authcfg = QString() ) const = 0 SIP_FACTORY;
```

Comment:
I am wondering if it doesn't make things over-complicated to use. Would it be easier to first create a `QgsExternalStorageStoredContent` and then have something like
`virtual ContentStatus store( QgsExternalStorageStoredContent *content )` ?